File: cpp/hard/41.first-missing-positive.cpp
/*
 * @lc app=leetcode id=41 lang=cpp
 *
 * [41] First Missing Positive
 */

#include <bits/stdc++.h>

using namespace std;

// @lc code=start
class Solution
{
public:
    int firstMissingPositive(vector<int> &nums)
    {
        int n = nums.size();
        int i = 0;

        while (i < n)
        {
            if (nums[i] > 0 && nums[i] <= n && nums[i] != nums[nums[i] - 1])
            {
                swap(nums[i], nums[nums[i] - 1]);
            }
            else
            {
                i++;
            }
        }

        for (int i = 0; i < n; i++)
        {
            if (nums[i] != i + 1)
            {
                return i + 1;
            }
        }

        return n + 1;
    }
};
// @lc code=end

File: rust/src/hard/41.first-missing-positive.rs
/*
 * @lc app=leetcode id=41 lang=rust
 *
 * [41] First Missing Positive
 */

// @lc code=start
impl Solution {
    pub fn first_missing_positive(mut nums: Vec<i32>) -> i32 {
        let mut i = 0;
        let n = nums.len();

        while i < n {
            if nums[i] > 0 && nums[i] <= n as i32 && nums[i] != nums[(nums[i] - 1) as usize] {
                let mut j = (nums[i] - 1) as usize;
                nums.swap(i, j);
            } else {
                i += 1;
            }
        }

        for i in 0..n {
            if (nums[i] != (i + 1) as i32) {
                return i as i32 + 1;
            }
        }

        return n as i32 + 1;
    }
}
// @lc code=end
